polyA/calculate_score.py

```diff
@@ -70,5 +74,81 @@ def calculate_score(
                 chunk_score += gap_init
         else:
             chunk_score += sub_matrix[seq1_char + seq2_char]
-
+            if char_complexity_adjustment:
+                chunk_score += char_complexity_adjustment[seq2_char]
     return chunk_score
+
+
+def calc_target_char_counts(
+    query_seq: str,
+    target_seq: str,
+):
+    query_char_counts = {"A": 0, "C": 0, "G": 0, "T": 0}
+    other_chars = set()
+    total_chars = 0
+    for q_base, t_base in zip(query_seq, target_seq):
+        if t_base not in ["A", "G", "T", "C"]:
+            other_chars.add(t_base)
+            continue
+        if q_base == "-":
+            continue
+        query_char_counts[t_base] += 1
+        total_chars += 1
+    return query_char_counts, total_chars, other_chars
+
+
+def calculate_complexity_adjusted_score(
+    char_background_freqs: Optional[Dict[str, float]],
+    query_seq: str,
+    target_seq: str,
+    lamb: float,
+):
+    char_complexity_adjustments: Dict[str, float] = {}
+    if char_background_freqs is None:
+        # set the complexity adjustment to zero for every char
+        for char in target_seq:
+            char_complexity_adjustments[char] = 0
+            if char == ".":
+                break
+        return char_complexity_adjustments
+
+    t_factor: float = 0
+    t_sum: float = 0
+    t_counts: int = 0
+    target_char_counts, total_chars, other_chars = calc_target_char_counts(
+        query_seq, target_seq
+    )
+
+    char_complexity_adjustments = {"A": 0, "C": 0, "G": 0, "T": 0}
+    for char, freq in char_background_freqs.items():
+        count = target_char_counts[char]
+        if count > 0 and freq > 0 and math.log(freq) != 0:
+            t_factor += count * math.log(count)
+            t_sum += count * math.log(freq)
+            t_counts += count
+            char_complexity_adjustments[char] += count * math.log(freq)
+
+    # char percent contribution to t_sum
+    for c, v in char_complexity_adjustments.items():
+        char_complexity_adjustments[c] /= t_sum
+
+    if t_counts != 0:
+        t_factor -= t_counts * math.log(t_counts)
+    t_sum -= t_factor
+
+    # char value contribution to t_sum
+    for c, v in char_complexity_adjustments.items():
+        char_complexity_adjustments[c] *= t_sum
+
+    # per position char value contribution to score adjustment
+    for c, v in char_complexity_adjustments.items():
+        if v != 0:
+            char_complexity_adjustments[c] /= target_char_counts[c]
+            char_complexity_adjustments[c] /= lamb
+            char_complexity_adjustments[c] += 0.999 / total_chars
```

Comment:
Where does `0.999` come from? It should probably be a constant in `constants.py`, along with documentation explaining what it is and why it has the value it has.